File: dfsbfs/puzzle.py
from collections import deque


def encode(x):
    return "".join(map(str, x[0])) + "".join(map(str, x[1])) + "".join(map(str, x[2]))

def decode(s):
    cur = []
    for i in range(0, 9, 3):
        cur.append(list(map(int, s[i:i+3])))
    return cur

valid = lambda i, j: 0<=i<3 and 0<=j<3

def solve():
    a = [list(map(int, input().split())) for _ in range(3)]
    q = deque()
    q.append((0, encode(a)))
    vis = set()
    ans = 0
    dd = [1, -1, 0, 0]
    while q:
        pl, cur = q.popleft()
        if cur == "123456780":
            ans = pl
            break
        nxt = decode(cur)
        if cur in vis: continue
        vis.add(cur)
        i, j = [(i, j) for i in range(3) for j in range(3) if nxt[i][j] == 0][0]
        for k in range(4):
            di, dj = dd[k], dd[-k-1]
            if valid(i+di, j+dj):
                nxt[i][j], nxt[i+di][j+dj] = nxt[i+di][j+dj], nxt[i][j]
                enc = encode(nxt)
                if enc in vis:
                    nxt[i][j], nxt[i+di][j+dj] = nxt[i+di][j+dj], nxt[i][j]
                    continue
                q.append((pl + 1, enc))
                nxt[i][j], nxt[i+di][j+dj] = nxt[i+di][j+dj], nxt[i][j]

    print(ans)




n = int(input())
for _ in range(n): solve()

File: dfsbfs/puzzle.cpp
#include <bits/stdc++.h>
using namespace std;

string encode(const vector<vector<int>>& x) {
    string result = "";
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            result += to_string(x[i][j]);
        }
    }
    return result;
}

vector<vector<int>> decode(const string& s) {
    vector<vector<int>> result(3, vector<int>(3));
    int idx = 0;
    for (int i = 0; i < 9; i += 3) {
        for (int j = 0; j < 3; ++j) {
            result[idx][j] = s[i + j] - '0';
        }
        ++idx;
    }
    return result;
}

bool valid(int i, int j) {
    return i >= 0 && i < 3 && j >= 0 && j < 3;
}

map<string, int> memo;

void solve() {
    vector<vector<int>> a(3, vector<int>(3));
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            cin >> a[i][j];
        }
    }
    
    cout << memo[encode(a)] << endl;
}

void precomp() {
    vector<vector<int>> a(3, vector<int>(3));
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            a[i][j] = i * 3 + j + 1;
        }
    }
    a[2][2] = 0;

    deque<pair<int, string>> q;
    memo[encode(a)] = 0;
    q.push_back({0, encode(a)});

    int ans = -1;
    vector<int> dd = {1, -1, 0, 0};
    
    while (!q.empty()) {
        int pl = q.front().first;
        string cur = q.front().second;
        q.pop_front();
        
        vector<vector<int>> nxt = decode(cur);
        
        int i, j;
        for (i = 0; i < 3; ++i) {
            for (j = 0; j < 3; ++j) {
                if (nxt[i][j] == 0) {
                    break;
                }
            }
            if (j < 3) break;
        }
        
        for (int k = 0; k < 4; ++k) {
            int di = dd[k], dj = dd[3 - k];
            if (valid(i + di, j + dj)) {
                swap(nxt[i][j], nxt[i + di][j + dj]);
                
                string enc = encode(nxt);
                if (memo.count(enc) == 0) {
                    q.push_back({pl + 1, enc});
                    memo[enc] = pl + 1;
                }
                
                swap(nxt[i][j], nxt[i + di][j + dj]);
            }
        }
    }
}

int main() {
    int n;
    cin >> n;

    precomp();
    
    for (int i = 0; i < n; ++i) {
        solve();
    }
    
    return 0;
}
